fix(requires-writer): use `@reach/router` path ranking algorithm to sort matchPaths

pieh opened the pull request:
(this is draft, not tested outside of checking if `requires-writer.js` unit tests pass

TO-DO:
- [x] add test cases to e2e tests and not only unit tests, so behaviour is tested (and not just implementation).

I've added few "integration style" tests there:
 - `will find static path before dynamic paths`
 - `will find path with dynamic paramter before path with wildcard`
 - `it will find most specific path with wildcard`

Fixes #19043

pieh commented on packages/gatsby/src/bootstrap/__tests__/requires-writer.js:
```diff
@@ -148,6 +149,10 @@ describe(`requires-writer`, () => {
+        {
+          path: `/another-custom-404`,
+          matchPath: `/*`,
+        },
```
(to make sure order of created pages doesn't impact final order of sorted pages)

pieh commented on packages/gatsby/src/bootstrap/requires-writer.js:
```diff
@@ -80,20 +102,15 @@ const getMatchPaths = pages => {
-      return b.index - a.index
+      // if specificity is the same we do lexigraphic comparison of path to ensure
+      // deterministic order regardless of order pages where created
+      return a.matchPath.localeCompare(b.matchPath)
```
This also need some eyes on (this is not exactly something needed for this fix, it just should mean that `matchPaths.json` order is always the same regardless of order of pages)

sever1an commented:
Is this the function we would ask reach-router to export? Should we open an issue there?
https://github.com/reach/router/blob/28a79e7fc3a3487cb3304210dc3501efb8a50eba/src/lib/utils.js#L228

pieh commented:
> Is this the function we would ask reach-router to export? Should we open an issue there?
> [reach/router:src/lib/utils.js@`28a79e7`#L228](https://github.com/reach/router/blob/28a79e7fc3a3487cb3304210dc3501efb8a50eba/src/lib/utils.js#L228)

I did modify the function a little bit (I'm passing just path instead of route object to avoid creating temporary objects - micro-optim), but essentially yes.

In my mind this shouldn't be a blocker for this PR - so this is opportunity for potential follow-up PRs (we don't need to hold off fix and wait for `@reach/router` to release new version - latest stable release was ~1 year ago, and they have committed more changes recently that are under `next` tag, so it might be a while for next stable version release)

sever1an commented:
I agree, it should be a followup pr/issue. We are already hosting a parallel implementation of route scoring, so there is no extra maintenance. 

Kind of thinking if we want it to happen before Gatsby 3.0.0, maybe start sooner than later....

pieh commented:
Added some e2e tests together with @madalynrose:

For reference:
e2e tests that I added have following results on:
 - current `latest`:
   ![Screenshot 2019-10-31 at 20 14 21](https://user-images.githubusercontent.com/419821/67979070-d470f580-fc1b-11e9-9c10-07c3b9d2e87c.png)
 - `gatsby@2.17.2` (before wildcard fix):
   ![Screenshot 2019-10-31 at 20 16 31](https://user-images.githubusercontent.com/419821/67979107-ebafe300-fc1b-11e9-8660-1138036fbcde.png)

wardpeet commented on yarn.lock:
```diff
@@ -6926,7 +6926,7 @@ csstype@^2.2.0, csstype@^2.5.7:
   integrity sha512-JsTaiksRsel5n7XwqPAfB0l3TFKdpjW/kgAELf9vrb5adGA7UCPLajKK5s3nFrcFm3Rkyp/Qkgl73ENc1UY3cA==
 
-"csstype@github:rtsao/csstype#c4e709f":
+csstype@rtsao/csstype#c4e709f:
```
I don't see any package.json changes,not saying this is bad. it's a little bit weird that's all :) (not a blocker)

pieh commented:
Published `gatsby@2.17.9`
